EIPS/eip-1559.md

```diff
@@ -32,34 +34,61 @@ Ethereum currently prices transaction fees using a simple auction mechanism, whe
-* Redefine the way the `tx.gasprice` field is used: define `tx.fee_premium = tx.gasprice // 2**128` and `tx.fee_cap = tx.gasprice % 2**128`
-* During transaction execution, we calculate the cost to the `tx.origin` and the gain to the `block.coinbase` as follows:
-  * Let `gasprice = min(BASEFEE + tx.fee_premium, tx.fee_cap)`. The `tx.origin` initially pays `gasprice * tx.gas`, and gets refunded `gasprice * (tx.gas - gasused)`.
-  * The `block.coinbase` gains `(gasprice - BASEFEE) * gasused`. If `gasprice < BASEFEE` (due to the `fee_cap`), this means that the `block.coinbase` _loses_ funds from this operation; in this case, check that the post-balance is non-negative and throw an exception if it is negative.
-As a default strategy, miners set `BASEFEE` as follows. Let `delta = block.gas_used - TARGET_GASUSED` (possibly negative). Set `BASEFEE = PARENT_BASEFEE + PARENT_BASEFEE * delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`, clamping this result inside of the allowable bounds if needed (with the parameter setting above clamping will not be required).
+For all blocks where `block.number >= INITIAL_FORK_BLKNUM`:
+
+For the gas limit:
+
+* `MAX_GAS_EIP1559` acts as the hard in-protocol gas limit, instead of the gas limit calculated using the previously existing formulas
+* The `GASLIMIT` field in the block header is the gas limit for the EIP1559 gas pool, and over the transition period this value increases until it reaches `MAX_GAS_EIP1559` at `FINAL_FORK_BLKNUM`
+* The gas limit for the legacy gas pool is `MAX_GAS_EIP1559 - GASLIMIT`, as `GASLIMIT` increases towards `MAX_GAS_EIP1559` gas is moved from the legacy pool into the EIP1559 pool until all of the gas is in the EIP1559 pool
+* At `block.number == INITIAL_FORK_BLKNUM`, let `GASLIMIT = (MAX_GAS_EIP1559 / 2)` so that the gas maximum is split evenly between the legacy and EIP1559 gas pools
+* As `block.number` increases towards `FINAL_FORK_BLKNUM`, at every block we shift `EIP1559_GAS_INCREMENT_AMOUNT` from the legacy pool into the EIP1559 gas pool
+* At `block.number >= FINAL_FORK_BLKNUM` the entire `MAX_GAS_EIP1559` is assigned to the EIP1559 gas pool and the legacy pool is empty
+* We enforce a per-transaction gas limit of 8,000,000.``
+
+For the gas price:
+
+* We add a new field to the block header, `BASEFEE`
+  * `BASEFEE` is maintained under consensus by the ethash engine
+* At `block.number == INITIAL_FORK_BLKNUM` we set `BASEFEE = INITIAL_BASEFEE`
+* As a default strategy, miners set `BASEFEE` as follows.
+  * Let `delta = block.gas_used - TARGET_GASUSED` (possibly negative).
+  * Set `BASEFEE = PARENT_BASEFEE + PARENT_BASEFEE * delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
+  * Clamp the resulting `BASEFEE` inside of the allowable bounds if needed, where a valid `BASEFEE` is one such that `abs(BASEFEE - PARENT_BASEFEE) <= max(1, PARENT_BASEFEE // BASEFEE_MAX_CHANGE_DENOMINATOR)`
```

Comment:
Are we letting miners adjust the basefee? I thought we are forcing the specific change in the protocol?

Reply:
I haven't had the time to thoroughly model the BASEFEE function, but in my limited experimentation there were scenarios where the specific change enforced by the protocol would result in a BASEFEE that falls out of the bounds described above. Only in that case is the result clamped within those bounds.

I will update all the references to "miners adjusting/setting BASEFEE" to reflect it being the protocol. I think this is a bit of semantic misunderstanding on my part, in my eyes the miners are still "setting" the BASEFEE in the blocks they mine, even if they have no control over what that set value is.

Reply:
Maybe to avoid simplicity reword this as:

* Let `gas_delta = block.gas_used - TARGET_GASUSED` (possibly negative).
* Let `basefee_delta = PARENT_BASEFEE * delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
* Let `max_basefee_delta = max(1, PARENT_BASEFEE // BASEFEE_MAX_CHANGE_DENOMINATOR)`
* Set `BASEFEE = PARENT_BASEFEE + max(-max_basefee_delta, min(max_basefee_delta, basefee_delta))`

Reply:
Actually, it's also important to specify cleanly what we mean by // in the negative case. Is `-7 // 5` equal to `-1` or `-2`? In python it's the latter, in some languages it's the former. Perhaps avoid negative numbers entirely?

* Let `max_basefee_delta = max(1, PARENT_BASEFEE // BASEFEE_MAX_CHANGE_DENOMINATOR)`
* If `block.gas_used >= TARGET_GASUSED`:
    * Let `gas_delta = block.gas_used - TARGET_GASUSED`
    * Let `basefee_delta = PARENT_BASEFEE * gas_delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
    * Set `BASEFEE = PARENT_BASEFEE + min(max_basefee_delta, basefee_delta)`
* `block.gas_used < TARGET_GASUSED`:
    * Let `gas_delta = TARGET_GASUSED - block.gas_used`
    * Let `basefee_delta = PARENT_BASEFEE * gas_delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`
    * Set `BASEFEE = max(0, PARENT_BASEFEE - min(max_basefee_delta, basefee_delta))`

Reply:
That's a great point, we are using `big.Int` division which is true Euclidean division where the remainder is bound above 0 so we round to negative infinity and would get `-2`. We can make that the specification, or avoid negative numbers entirely. What do you think is best?